### PR TITLE
Allow to delete schemas.

### PR DESCRIPTION
**Pull request by PawelPawelec-RDX**

### CODE REVIEW

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 1]
```diff
@@ -151,12 +173,48 @@ public void ProcessChanges()
                     EntityId = lookup.EntityId,
                     SchemaHash = entry.Key.SchemaHash.ConvertFromHex(),
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:174]: Question: shouldn't we check if given `entityId + schemaHash` doesn't exist already? Are we sure we do not have any duplicates in our database?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 2]
```diff
@@ -151,12 +173,48 @@ public void ProcessChanges()
...
+                {
+                    var currentPosition = aggregate.EntryIds.IndexOf(previousEntry.Id);
+
+                    if (currentPosition != -1)
+                    {
+                        aggregate.EntryIds.RemoveAt(currentPosition);
+                    }
+                    else
+                    {
+                        throw new UnreachableException($"Unexpected situation where SchemaEntryDefinition with EntityId:{entryLookup.EntityId}, SchemaHash:{deletedSchemaHash} got deleted but wasn't found in aggregate table.");
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:198]: Are we 100% sure schema cannot be deleted on its own somehow? Be advised schemas are not limited to package blueprints.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 3]
```diff
@@ -151,12 +173,48 @@ public void ProcessChanges()
...
+                    }
+                }
+                else
+                {
+                    throw new UnreachableException($"Unexpected situation where SchemaEntryDefinition with EntityId:{entryLookup.EntityId}, SchemaHash:{deletedSchemaHash} got deleted but wasn't found in gateway database.");
+                }
+
+                var entryDefinition = new SchemaEntryDefinition
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:206]: Ah damn... `*_definition` tables are not meant to keep track of history. They're meant to hold immutable information. 
1. Our queries that rely on `*_definition` tables do not use lateral joins. 
2. Are we sure it is OK to say "schema got deleted" from the network? Is it deletable entity at all? I guess we can say "schema is no longer used by XYZ" (i.e. aggregate) though? Guess we don't need to store `is_deleted` at all?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 4]
```diff
@@ -196,16 +254,51 @@ LIMIT 1
             e => e.EntityId);
     }
 
+    private async Task<IDictionary<SchemaDefinitionEntryDbLookup, SchemaEntryDefinition>> LoadSchemaDefinitionsToBeDeleted()
+    {
+        var lookupSet = new HashSet<SchemaDefinitionEntryDbLookup>();
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:259]: How about `var lookupSet = _changes.AsEnumerable().SelectMany(c => c.DeletedSchemaHashes).Select(x => new SchemaDefinitionEntryDbLookup(...)).ToHashSet()`?

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 5]
```diff
@@ -196,16 +254,51 @@ LIMIT 1
...
+        return await _context.ReadHelper.LoadDependencies<SchemaDefinitionEntryDbLookup, SchemaEntryDefinition>(
+            @$"
+WITH variables (entity_id, schema_hash) AS (
+    SELECT UNNEST({entityIds}), UNNEST({schemaHashes})
+)
+SELECT sedh.*
+FROM variables
+INNER JOIN LATERAL (
+    SELECT *
+    FROM schema_entry_definition
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:283]: As said previously this table is not meant for history lookups.

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs [hunk 6]
```diff
@@ -108,9 +115,24 @@ public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity refere
         }
     }
 
+    public void VisitDelete(CoreModel.SubstateId substateId, ReferencedEntity referencedEntity, long stateVersion)
+    {
+        if (substateId.SubstateType == CoreModel.SubstateType.SchemaEntry)
+        {
+            var keyHex = ((CoreModel.MapSubstateKey)substateId.SubstateKey).KeyHex;
+            var schemaHash = ScryptoSborUtils.DataToProgrammaticScryptoSborValueBytes(keyHex.ConvertFromHex(), _networkId);
```
Comment (krzlabrdx) [inline src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:123]: Emm... I guess the key itself isn't equal to schema hash hex-encoded?

**Review by krzlabrdx — CHANGES_REQUESTED** [review]

File: src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs (hunk 6 above)
Comment (PawelPawelec-RDX) [reply src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/EntitySchemaProcessor.cs:123]: yes, it isn't

**Review by krzlabrdx — APPROVED** [review]

**Pull request merged**